bazel/ray_deps_setup.bzl

```diff
@@ -1,128 +1,198 @@
-    RULES_JVM_EXTERNAL_TAG = "2.10"
-    RULES_JVM_EXTERNAL_SHA = "1bbf2e48d07686707dd85357e9a94da775e1dbd7c464272b3664283c9c716d26"
+def github_repository(**kwargs):
+    """
+    Conveniently chooses between ZIP, git, etc. GitHub repositories.
+    We prefer ZIP downloads, since they're smaller and faster
+    due to the lack of history.
+
+    The remote parameter is required.
+    One of {commit, tag, branch} is also required.
+    The name parameter is optional but recommended.
+    Set sha256 = True if the ZIP archive hash is unknown, then update it ASAP
+    to allow caching & avoid repeated downloads on every build.
+
+    If build_file is a string, it refers to a custom build file.
+    If build_file is absent, the project is assumed to contain a build file.
+    If build_file == True, then the build file name is auto-deduced.
+
+    If path   is present , local repository is assumed.
+    If sha256 == <string>, obtains ZIP download with the given hash (fast)
+    If sha256 == True    , obtains ZIP download without knowing the hash (fast)
+    If sha256 is absent  , obtains clone via git (slow and NOT recommended)
+
+    strip_prefix is automatic by default.
+    Set strip_prefix=<string> (or False) to override.
+
+    Some common parameters (like shallow_since)
+    are automatically stripped when not applicable.
+    """
+
+    remote = kwargs.pop("remote")  # required
+
+    GIT_SUFFIX = ".git"  # We need to strip this
+    if remote.endswith(GIT_SUFFIX):
+        remote = remote[:len(remote) - len(GIT_SUFFIX)]
+    project = remote.split("//", 1)[1].split("/")[2]
+
+    commit = kwargs.pop("commit", None)
+    tag = kwargs.pop("tag", None)
+    branch = kwargs.pop("branch", None)
+    treeish = commit or tag or branch
+    if not treeish: fail("Missing commit, tag, or branch argument")
+    build_file = kwargs.pop("build_file", None)
+    url = kwargs.pop("url", "%s/archive/%s.zip" % (remote, treeish))
+    sha256 = kwargs.pop("sha256", None)
+    shallow_since = kwargs.pop("shallow_since", None)
+    strip_prefix = kwargs.pop("strip_prefix", None)
+    kwargs.setdefault("name", project.replace("-", "_"))
+
+    if build_file == True:
+        # Auto-generate build file name
+        kwargs.setdefault(
+            "build_file",
+            "@//%s:%s" % ("bazel", "BUILD." + kwargs["name"]))
+
+    if "path" in kwargs:
+        native.local_repository(**kwargs)
+    else:
+        if strip_prefix in (True, None):
+            kwargs.setdefault("strip_prefix", "%s-%s" % (project, treeish))
+        if not sha256:
+            # git clone
+            if commit != None: kwargs.setdefault("commit", commit)
+            if tag != None: kwargs.setdefault("tag", tag)
+            if branch != None: kwargs.setdefault("branch", branch)
+            kwargs.setdefault("shallow_since", shallow_since)
+            kwargs.setdefault("remote", remote)
+            if "build_file" in kwargs:
+                new_git_repository(**kwargs)
```

Comment:
please just call this and `git_repository` below with expanded keyword arguments for clarity, otherwise I have to trace back all of the code that modifies `kwargs` to understand what this line is passing in

Reply:
and `http_archive` below

Reply:
Hmm okay, if you confirm https://github.com/ray-project/ray/pull/6193/files#r348190605 then I'll do this.

Reply:
Update: I think this isn't workable anymore in light of https://github.com/ray-project/ray/pull/6193#discussion_r348218943.